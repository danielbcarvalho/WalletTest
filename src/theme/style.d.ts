import 'styled-components/native';

declare module 'styled-components/native' {
  interface TipographyKindProps {
    fontFamily: string;
    fontSize: string;
    lineHeight?: string;
  }

  interface TipographyProps {
    h1: TipographyKindProps;
    h2: TipographyKindProps;
    h3: TipographyKindProps;
    h4: TipographyKindProps;
    h5: TipographyKindProps;
    p: TipographyKindProps;
    pSmall: TipographyKindProps;
  }

  interface LabelProps extends TipographyKindProps {
    color: string;
  }
  interface InputProps extends TipographyKindProps {
    color: string;
    backgroundColor: string;
    borderRadius: string;
    padding: string;
    label: LabelProps;
  }

  interface ButtonProps extends TipographyKindProps {
    radius: string;
    primary: {
      backgroundColor: string;
      color: string;
    };
    secondary: {
      backgroundColor: string;
      color: string;
    };
    disabled: {
      backgroundColor: string;
      color: string;
    };
  }

  interface PalleteProps {
    primary: {
      main: string;
    };
    success: {
      main: string;
    };
    alert: {
      main: string;
    };
    warning: {
      main: string;
    };
    info: {
      main: string;
    };
    background: {
      default: string;
      overlay: string;
    };
    textOnBackground: {
      default: string;
      overlay: string;
    };
  }

  interface Header {
    title: HeaderTitle;
  }

  interface HeaderTitle extends TipographyKindProps {
    register: {
      color: string;
    };
    cardList: {
      color: string;
    };
  }

  export interface DefaultTheme {
    pallete: PalleteProps;
    typography: TipographyProps;
    components: {
      button: ButtonProps;
      input: InputProps;
      header: Header;
    };
  }
}
