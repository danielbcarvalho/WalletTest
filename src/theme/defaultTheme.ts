import { DefaultTheme } from 'styled-components/native';

export const defaultTheme: DefaultTheme = {
  pallete: {
    primary: {
      main: '#142995',
    },
    success: {
      main: '#A5FF32',
    },
    alert: {
      main: '#FF0000',
    },
    warning: {
      main: '#FFB800',
    },
    info: {
      main: '#12C2E9',
    },
    background: {
      default: '#142995',
      overlay: '#EEEEEE',
    },
    textOnBackground: {
      default: '#FFF',
      overlay: '#142995',
    },
  },
  typography: {
    h1: {
      fontFamily: 'PTSans-Caption',
      fontSize: '28px',
      lineHeight: '42px',
    },
    h2: {
      fontFamily: 'PTSans-Caption',
      fontSize: '26px',
      lineHeight: '38px',
    },
    h3: {
      fontFamily: 'PTSansCaption',
      fontSize: '22px',
      lineHeight: '34px',
    },
    h4: {
      fontFamily: 'PTSans-Regular',
      fontSize: '20px',
      lineHeight: '32px',
    },
    h5: {
      fontFamily: 'PTSans-Regular',
      fontSize: '18px',
      lineHeight: '30px',
    },
    p: {
      fontFamily: 'PTSans-Regular',
      fontSize: '16px',
      lineHeight: '28px',
    },
    pSmall: {
      fontFamily: 'PTSans-Regular',
      fontSize: '14px',
      lineHeight: '26px',
    },
  },
  components: {
    button: {
      radius: '12px',
      fontFamily: 'PTSans-Regular',
      fontSize: '18px',
      lineHeight: '28px',
      primary: {
        backgroundColor: '#12C2E9',
        color: '#FFFFFF',
      },
      secondary: {
        backgroundColor: '#A5FF32',
        color: '#142995',
      },
      disabled: {
        backgroundColor: '#EEEEEE',
        color: '#BBBBBB',
      },
    },
    input: {
      label: {
        fontFamily: 'PTSans-Regular',
        fontSize: '14px',
        color: '#FFFFFF',
        lineHeight: '16px',
      },
      fontFamily: 'PTSans-Regular',
      fontSize: '16px',
      color: '#111',
      backgroundColor: '#EEEEEE',
      borderRadius: '12px',
      padding: '16px',
    },
    header: {
      title: {
        fontFamily: 'PTSans-Caption',
        fontSize: '22px',
        lineHeight: '24px',
        register: {
          color: '#142995',
          backgroundColor: '#FFFFFF',
        },
        cardList: {
          color: '#12C2E9',
        },
      },
    },
    card: {
      black: {
        backgroundColor: '#000000',
        color: '#FFFFFF',
      },
      green: {
        backgroundColor: '#A5FF32',
        color: '#3F3F3F',
      },
      title: {
        fontFamily: 'PTSans-Regular',
        fontSize: '18px',
        lineHeight: '20px',
      },
      name: {
        fontFamily: 'PTSans-Regular',
        fontSize: '16px',
        lineHeight: '18px',
      },
      text: {
        fontFamily: 'PTSans-Regular',
        fontSize: '14px',
        lineHeight: '16px',
      },
    },
  },
};

export const mockDefaultTheme: object = defaultTheme;
