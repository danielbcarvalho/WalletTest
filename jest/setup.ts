import 'react-native-gesture-handler/jestSetup';
import { mockDefaultTheme } from '../src/theme/defaultTheme';

jest.mock('styled-components', () => {
  const original = jest.requireActual('styled-components');

  return {
    ...original,
    useTheme: () => {
      return mockDefaultTheme;
    },
  };
});
