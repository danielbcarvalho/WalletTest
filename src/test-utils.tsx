// TODO: Verificar essa melhoria nos testes
import React from 'react';
import {
  render,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { mockDefaultTheme } from './theme/defaultTheme';

interface IAllThemeProviders {
  children: React.ReactElement;
}
const AllTheProviders = ({ children }: IAllThemeProviders) => {
  return <ThemeProvider theme={mockDefaultTheme}>{children}</ThemeProvider>;
};

function customRender(
  ui: React.ReactElement,
  options = {} as RenderOptions,
): RenderAPI {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export function rendererCreate(
  component: React.ReactElement,
): ReactTestRenderer {
  return renderer.create(<AllTheProviders>{component}</AllTheProviders>);
}

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
