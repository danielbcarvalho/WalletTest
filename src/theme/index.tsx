import React, { createContext, useMemo } from 'react';
import { ThemeProvider as ThemeProviderStyled } from 'styled-components/native';

import { defaultTheme } from './defaultTheme';

export enum Theme {
  light = 'light',
  dark = 'dark',
}

interface IApplication {
  children: JSX.Element;
}

const themes = {
  light: defaultTheme,
};

export const ThemeContext = createContext({
  theme: Theme.light,
});

export function ThemeProvider({ children }: IApplication) {
  const theme = Theme.light;

  const memoizedValue = useMemo(() => ({ theme }), [theme]);

  return (
    <ThemeContext.Provider value={memoizedValue}>
      <ThemeProviderStyled theme={themes[theme]}>
        {children}
      </ThemeProviderStyled>
    </ThemeContext.Provider>
  );
}
