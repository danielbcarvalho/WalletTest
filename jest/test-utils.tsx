import React, { ReactElement, FC } from 'react';
import { render } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ThemeProvider } from '../src/theme';

type Options = Parameters<typeof render>[1];

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: Infinity,
    },
  },
});
const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ThemeProvider>
);

const customRender = (ui: ReactElement, options?: Options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
