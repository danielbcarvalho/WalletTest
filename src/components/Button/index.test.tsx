import React from 'react';
import { fireEvent } from '@testing-library/react-native';

import { render } from '../../../jest/test-utils';
import { defaultTheme } from '../../theme/defaultTheme';

import { Button } from '.';

describe('Button component', () => {
  test('renders with default props', () => {
    const { getByText, getByTestId } = render(
      <Button title="Default Button" />,
    );
    const buttonTextElement = getByText('Default Button');
    expect(buttonTextElement).toBeDefined();

    const buttonElement = getByTestId('multipurpose-button');

    expect(buttonElement.props.style.backgroundColor).toBe(
      defaultTheme.components.button.primary.backgroundColor,
    );
  });

  test('renders with primary type', () => {
    const { getByText, getByTestId } = render(
      <Button title="Primary Button" type="primary" />,
    );
    const buttonTextElement = getByText('Primary Button');
    expect(buttonTextElement).toBeDefined();

    const buttonElement = getByTestId('multipurpose-button');

    expect(buttonElement.props.style.backgroundColor).toBe(
      defaultTheme.components.button.primary.backgroundColor,
    );
  });

  test('renders with secondary type', () => {
    const { getByText, getByTestId } = render(
      <Button title="Secondary Button" type="secondary" />,
    );
    const buttonTextElement = getByText('Secondary Button');
    expect(buttonTextElement).toBeDefined();

    const buttonElement = getByTestId('multipurpose-button');

    expect(buttonElement.props.style.backgroundColor).toBe(
      defaultTheme.components.button.secondary.backgroundColor,
    );
  });

  test('renders disabled', () => {
    const { getByText, getByTestId } = render(
      <Button title="Disabled Button" disabled />,
    );
    const buttonTextElement = getByText('Disabled Button');
    expect(buttonTextElement).toBeDefined();

    const buttonElement = getByTestId('multipurpose-button');

    expect(buttonElement.props.style.backgroundColor).toBe(
      defaultTheme.components.button.disabled.backgroundColor,
    );
  });

  test('fires onPress event when pressed', () => {
    const onPressMock = jest.fn();

    const { getByText } = render(
      <Button title="Clickable Button" onPress={onPressMock} />,
    );
    const buttonElement = getByText('Clickable Button');
    expect(buttonElement).toBeDefined();

    fireEvent.press(buttonElement);

    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  test('matches snapshot', () => {
    const { toJSON } = render(<Button title="Default Button" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
