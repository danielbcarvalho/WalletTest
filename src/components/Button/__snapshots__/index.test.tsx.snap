// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#12C2E9",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "justifyContent": "center",
      "marginBottom": 10,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 10,
      "maxHeight": 56,
      "minHeight": 56,
      "opacity": 1,
      "paddingBottom": 10,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 10,
    }
  }
  testID="multipurpose-button"
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "PTSans-Regular",
        "fontSize": 18,
      }
    }
    type="primary"
  >
    Default Button
  </Text>
</View>
`;
